File: app/src/main/java/de/davis/passwordmanager/ui/dashboard/DashboardFragment.java
package de.davis.passwordmanager.ui.dashboard;

import android.content.Context;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.core.view.MenuProvider;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.RecyclerView;
import androidx.slidingpanelayout.widget.SlidingPaneLayout;

import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.color.MaterialColors;

import de.davis.passwordmanager.R;
import de.davis.passwordmanager.dashboard.DashboardAdapter;
import de.davis.passwordmanager.dashboard.viewholders.BasicViewHolder;
import de.davis.passwordmanager.databinding.FragmentDashboardBinding;
import de.davis.passwordmanager.manager.ActivityResultManager;
import de.davis.passwordmanager.security.element.SecureElementDetail;
import de.davis.passwordmanager.security.element.SecureElementManager;
import de.davis.passwordmanager.ui.callbacks.SearchViewBackPressedHandler;
import de.davis.passwordmanager.ui.callbacks.SlidingBackPaneManager;
import de.davis.passwordmanager.ui.viewmodels.DashboardViewModel;
import de.davis.passwordmanager.ui.viewmodels.ScrollingViewModel;
import de.davis.passwordmanager.ui.views.AddBottomSheet;
import de.davis.passwordmanager.ui.views.FilterBottomSheet;
import de.davis.passwordmanager.ui.views.OptionBottomSheet;

public class DashboardFragment extends Fragment implements SearchView.OnQueryTextListener {

    private FragmentDashboardBinding binding;

    private DashboardViewModel viewModel;
    private ScrollingViewModel scrollingViewModel;

    private boolean oldState = true;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentDashboardBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding.listPane.recyclerView.setHasFixedSize(true);

        NavHostFragment navHostFragment = (NavHostFragment) getChildFragmentManager().findFragmentById(R.id.elementContainer);
        if(navHostFragment == null)
            return;

        NavController navController = navHostFragment.getNavController();
        binding.getRoot().setLockMode(SlidingPaneLayout.LOCK_MODE_LOCKED);

        ((AppCompatActivity)requireActivity()).setSupportActionBar(binding.listPane.searchBar);

        binding.listPane.viewAddFirst.setOnClickListener(v -> showBottomSheet());

        ActivityResultManager arm = ActivityResultManager.getOrCreateManager(getClass(), this);
        arm.registerCreate();
        arm.registerEdit(null);

        SecureElementManager manager = SecureElementManager.createNew(sem -> {
            boolean hasElements = sem.hasElements();
            binding.listPane.progress.setVisibility(View.GONE);

            binding.listPane.recyclerView.setVisibility(hasElements ? View.VISIBLE : View.GONE);
            binding.listPane.viewToShow.setVisibility(hasElements ? View.GONE : View.VISIBLE);
        });

        DashboardAdapter dashboardAdapter = manager.getAdapter();
        dashboardAdapter.applyWithTracker(binding.listPane.recyclerView);

        BasicViewHolder.OnItemClickedListener onItemClickedListener = element -> {
            scrollingViewModel.setVisibility(false);
            binding.listPane.searchView.hide();
            Bundle bundle = new Bundle();
            bundle.putSerializable("element", element);
            navController.popBackStack();
            navController.navigate(SecureElementDetail.getFor(element).getViewFragmentId(), bundle);

            binding.getRoot().open();
        };

        dashboardAdapter.setOnItemClickedListener(onItemClickedListener);

        DashboardAdapter searchResultAdapter = new DashboardAdapter();
        searchResultAdapter.setOnItemClickedListener(onItemClickedListener);
        binding.listPane.recyclerViewResults.setAdapter(searchResultAdapter);


        addMenu(dashboardAdapter);

        binding.listPane.searchView.getEditText().addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {}

            @Override
            public void afterTextChanged(Editable s) {
                viewModel.search(s.toString());
            }
        });


        viewModel = new ViewModelProvider(this, ViewModelProvider.Factory.from(DashboardViewModel.initializer)).get(DashboardViewModel.class);
        viewModel.getElements().observe(getViewLifecycleOwner(), secureElements -> SecureElementManager.getInstance().update(secureElements));
        viewModel.getSearchResults().observe(getViewLifecycleOwner(), secureElements -> {
            searchResultAdapter.update(secureElements);
            searchResultAdapter.setFilter(viewModel.getSearchQuery());
            if(!TextUtils.isEmpty(viewModel.getSearchQuery()) && secureElements.isEmpty()){
                binding.listPane.noResults.setVisibility(View.VISIBLE);
                return;
            }

            binding.listPane.noResults.setVisibility(View.GONE);
        });


        scrollingViewModel = new ViewModelProvider(requireActivity()).get(ScrollingViewModel.class);

        SlidingBackPaneManager slidingBackPaneManager = new SlidingBackPaneManager(binding.slidingPaneLayout, scrollingViewModel);
        requireActivity().getOnBackPressedDispatcher().addCallback(getViewLifecycleOwner(), slidingBackPaneManager);
        requireActivity().getOnBackPressedDispatcher().addCallback(getViewLifecycleOwner(), new SearchViewBackPressedHandler(binding.listPane.searchView));


        //Animation for fab and bottom nav bar
        binding.listPane.recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                scrollingViewModel.setConsumedY(dy);
            }
        });
    }

    @Override
    public void onPause() {
        super.onPause();
        oldState = Boolean.TRUE.equals(scrollingViewModel.getVisibility().getValue());
        scrollingViewModel.setVisibility(false);
    }

    @Override
    public void onResume() {
        super.onResume();
        scrollingViewModel.setVisibility(oldState);
    }

    private void addMenu(DashboardAdapter dashboardAdapter){
        requireActivity().addMenuProvider(new MenuProvider() {
            @Override
            public void  onCreateMenu(@NonNull Menu menu, @NonNull MenuInflater menuInflater) {
                menuInflater.inflate(R.menu.view_menu, menu);
                dashboardAdapter.setStateChangeHandler(selectedItems -> {
                    requireActivity().invalidateMenu();
                    binding.listPane.searchBar.setHint(selectedItems > 0 ? getString(R.string.selected_items, selectedItems) : getString(android.R.string.search_go));
                });
            }

            @Override
            public void onPrepareMenu(@NonNull Menu menu) {
                MenuProvider.super.onPrepareMenu(menu);
                menu.findItem(R.id.more).setVisible(dashboardAdapter.getTracker().hasSelection());
            }

            @Override
            public boolean onMenuItemSelected(@NonNull MenuItem menuItem) {
                if(menuItem.getItemId() == R.id.more){
                    OptionBottomSheet optionBottomSheet = new OptionBottomSheet(requireContext(), null);
                    optionBottomSheet.show();
                }else if(menuItem.getItemId() == R.id.filter){
                    new FilterBottomSheet().show(getParentFragmentManager(), "FilterDialog");
                }
                return true;
            }
        }, getViewLifecycleOwner(), Lifecycle.State.RESUMED);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        SecureElementManager.getInstance().getAdapter().getTracker().onSaveInstanceState(outState);
        outState.putCharSequence("searchbar_hint", binding.listPane.searchBar.getHint());
    }

    @Override
    public void onViewStateRestored(@Nullable Bundle savedInstanceState) {
        super.onViewStateRestored(savedInstanceState);
        SecureElementManager.getInstance().getAdapter().getTracker().onRestoreInstanceState(savedInstanceState);
        if(savedInstanceState == null)
            return;

        binding.listPane.searchBar.setHint(savedInstanceState.getCharSequence("searchbar_hint", getString(android.R.string.search_go)));
    }

    private void showBottomSheet(){
        new AddBottomSheet().show(getParentFragmentManager(), "add-bottom-sheet");
    }

    @Override
    public boolean onQueryTextSubmit(String query) {
        return true;
    }

    @Override
    public boolean onQueryTextChange(String newText) {
        viewModel.search(newText);
        return true;
    }

    public static class ScrollingViewBehavior extends AppBarLayout.ScrollingViewBehavior {

        private boolean initialized = false;

        public ScrollingViewBehavior() {}

        public ScrollingViewBehavior(@NonNull Context context, @Nullable AttributeSet attrs) {
            super(context, attrs);
        }

        @Override
        public boolean onDependentViewChanged(
                @NonNull CoordinatorLayout parent, @NonNull View child, @NonNull View dependency) {
            boolean changed = super.onDependentViewChanged(parent, child, dependency);
            if (!initialized && dependency instanceof AppBarLayout appBarLayout) {
                initialized = true;
                setAppBarLayoutColor(appBarLayout);
            }
            return changed;
        }

        private void setAppBarLayoutColor(AppBarLayout appBarLayout) {
            appBarLayout.setBackgroundColor(MaterialColors.getColor(appBarLayout, com.google.android.material.R.attr.colorSurface));

            // Remove AppBarLayout elevation shadow
            appBarLayout.setElevation(0);
        }

        @Override
        protected boolean shouldHeaderOverlapScrollingChild() {
            return false;
        }
    }
}

File: app/src/main/java/de/davis/passwordmanager/ui/dashboard/DashboardFragment.kt
package de.davis.passwordmanager.ui.dashboard

import android.annotation.SuppressLint
import android.os.Bundle
import android.os.Parcelable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.os.bundleOf
import androidx.core.widget.doAfterTextChanged
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.navigation.NavController
import androidx.navigation.fragment.NavHostFragment
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.OnScrollListener
import androidx.slidingpanelayout.widget.SlidingPaneLayout
import androidx.slidingpanelayout.widget.SlidingPaneLayout.PanelSlideListener
import de.davis.passwordmanager.R
import de.davis.passwordmanager.database.ElementType
import de.davis.passwordmanager.database.dtos.Item
import de.davis.passwordmanager.database.dtos.SecureElement
import de.davis.passwordmanager.database.dtos.TagWithCount
import de.davis.passwordmanager.databinding.FragmentDashboardBinding
import de.davis.passwordmanager.filter.Filter
import de.davis.passwordmanager.ktx.doFlowInLifecycle
import de.davis.passwordmanager.ktx.getParcelableCompat
import de.davis.passwordmanager.manager.ActivityResultManager
import de.davis.passwordmanager.ui.callbacks.SearchViewBackPressedHandler
import de.davis.passwordmanager.ui.callbacks.SlidingBackPaneManager
import de.davis.passwordmanager.ui.dashboard.managers.AbsItemManager
import de.davis.passwordmanager.ui.dashboard.managers.ElementItemManager
import de.davis.passwordmanager.ui.dashboard.managers.TagItemManager
import de.davis.passwordmanager.ui.dashboard.menuprovider.DefaultElementMenuProvider
import de.davis.passwordmanager.ui.viewmodels.ScrollingViewModel
import kotlinx.coroutines.flow.collectLatest
import kotlinx.parcelize.Parcelize

class DashboardFragment : Fragment() {

    private lateinit var binding: FragmentDashboardBinding

    private val adapter = DashboardAdapter { updateUI() }

    private val scrollingViewModel: ScrollingViewModel by activityViewModels()
    private val dashboardViewModel: DashboardViewModel by viewModels()

    private lateinit var navController: NavController

    private lateinit var dashboardHandler: DashboardHandler

    private val onElementClicked = { e: SecureElement -> dashboardHandler.onElementClicked(e) }

    @Parcelize
    data class DashboardHandler(
        val hideElementTypes: List<ElementType> = emptyList(),
        val onElementClicked: (SecureElement) -> Unit
    ) : Parcelable

    private val defaultHandler = DashboardHandler { launchElement(it) }


    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentDashboardBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        dashboardHandler = arguments?.getParcelableCompat(
            EXTRA_DASHBOARD_HANDLER,
            DashboardHandler::class.java
        ) ?: defaultHandler
    }

    @SuppressLint("PrivateResource")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        dashboardViewModel.initiateState(dashboardHandler.hideElementTypes)

        binding.root.lockMode = SlidingPaneLayout.LOCK_MODE_LOCKED

        val menuProvider = DefaultElementMenuProvider(parentFragmentManager, adapter)
        requireActivity().addMenuProvider(
            menuProvider,
            viewLifecycleOwner,
            Lifecycle.State.STARTED
        )

        adapter.applyRecyclerView(
            binding.listPane.recyclerView,
            enableSelection = dashboardHandler == defaultHandler
        ) {
            binding.listPane.searchBar.hint = if (it.isNotEmpty())
                getString(R.string.selected_items, it.size)
            else
                getString(android.R.string.search_go)

            menuProvider.updateMenu(requireActivity()) {
                selectedElements = it
            }
        }

        val navHostFragment =
            childFragmentManager.findFragmentById(R.id.elementContainer) as NavHostFragment
        navController = navHostFragment.navController

        (requireActivity() as AppCompatActivity).setSupportActionBar(binding.listPane.searchBar)


        val backToTagsCallback = object : OnBackPressedCallback(false) {
            override fun handleOnBackPressed() {
                dashboardViewModel.updateState(ListState.Tag)
            }
        }

        val slidingBackPaneManager =
            SlidingBackPaneManager(binding.slidingPaneLayout, scrollingViewModel)

        binding.slidingPaneLayout.addPanelSlideListener(object : PanelSlideListener {
            override fun onPanelSlide(panel: View, slideOffset: Float) {}

            override fun onPanelOpened(panel: View) {
                backToTagsCallback.isEnabled = false
            }

            override fun onPanelClosed(panel: View) {
                backToTagsCallback.isEnabled = true
            }

        })

        requireActivity().onBackPressedDispatcher.apply {
            addCallback(
                viewLifecycleOwner,
                slidingBackPaneManager
            )

            addCallback(
                viewLifecycleOwner,
                SearchViewBackPressedHandler(binding.listPane.searchView)
            )

            addCallback(viewLifecycleOwner, backToTagsCallback)
        }

        // Animation for fab and bottom nav bar
        binding.listPane.recyclerView.addOnScrollListener(object : OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                scrollingViewModel.setConsumedY(dy)
            }
        })
        // ------------------------------------

        ActivityResultManager.getOrCreateManager(javaClass, this).apply {
            registerCreate()
            registerEdit {}
        }

        viewLifecycleOwner.doFlowInLifecycle(dashboardViewModel.state) {
            collectLatest { pair ->
                @Suppress("UNCHECKED_CAST")
                pair?.let { (listState, data) ->
                    // handle menu
                    menuProvider.updateMenu(requireActivity()) {
                        filterVisible = listState !is ListState.Tag
                    }
                    // -----------

                    val absItemManager = when (listState) {
                        is ListState.Tag -> TagItemManager(
                            data.data as List<TagWithCount>,
                            onClick = {
                                dashboardViewModel.updateState(ListState.Element(it.tag.name))
                            }
                        )

                        // Element and AllElements
                        else -> {
                            val elements = data.data as List<SecureElement>
                            if (data is ListData.Elements) {
                                Filter.updateTags(data.tags)
                            }

                            ElementItemManager(
                                elements,
                                onElementClicked,
                                childFragmentManager
                            )
                        }
                    }

                    update(absItemManager)

                    // Update icon
                    backToTagsCallback.isEnabled =
                        listState is ListState.Element && !binding.slidingPaneLayout.isOpen

                    if (backToTagsCallback.isEnabled) {
                        binding.listPane.searchBar.navigationIcon =
                            AppCompatResources.getDrawable(
                                requireContext(),
                                R.drawable.ic_baseline_close_24
                            )

                        binding.listPane.searchBar.setNavigationOnClickListener {
                            dashboardViewModel.updateState(ListState.Tag)
                        }
                    } else {
                        binding.listPane.searchBar.navigationIcon =
                            AppCompatResources.getDrawable(
                                requireContext(),
                                com.google.android.material.R.drawable.ic_search_black_24
                            )

                        binding.listPane.searchBar.setNavigationOnClickListener(null)
                    }
                }
            }
        }

        binding.listPane.searchView.editText.doAfterTextChanged {
            dashboardViewModel.search(it.toString())
        }
        val searchResultAdapter = DashboardAdapter {}
        searchResultAdapter.applyRecyclerView(binding.listPane.recyclerViewResults, false)

        doFlowInLifecycle(dashboardViewModel.searchResults) {
            collectLatest {
                searchResultAdapter.update(
                    ElementItemManager(
                        it.second,
                        onElementClicked,
                        childFragmentManager
                    )
                )
                searchResultAdapter.filter = it.first

                binding.listPane.noResultsSearch.visibility =
                    if (it.first.isNotEmpty() && it.second.isEmpty()) View.VISIBLE
                    else View.GONE
            }
        }


        arguments?.getParcelableCompat("element", SecureElement::class.java)?.let {
            onElementClicked(it)
        }
    }

    private fun launchElement(element: SecureElement) {
        scrollingViewModel.setFabVisibility(false)
        binding.listPane.searchView.hide()
        val bundle = bundleOf("element" to element)
        navController.apply {
            popBackStack()
            navigate(element.elementType.viewFragmentId, bundle)
        }

        binding.root.open()
    }

    override fun onPause() {
        super.onPause()
        scrollingViewModel.setFabVisibility(false)
    }

    override fun onResume() {
        super.onResume()
        scrollingViewModel.setFabVisibility(true)
    }

    private fun update(update: AbsItemManager<out Item>) {
        adapter.update(update)

        updateUI()
    }

    private fun updateUI() {
        binding.listPane.progress.visibility = View.GONE

        binding.listPane.noResults.visibility =
            if (adapter.itemCount > 0) View.GONE else View.VISIBLE
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        adapter.onSaveInstanceState(outState)
    }

    override fun onViewStateRestored(savedInstanceState: Bundle?) {
        super.onViewStateRestored(savedInstanceState)
        adapter.onRestoreInstanceState(savedInstanceState)
    }

    companion object {
        const val EXTRA_DASHBOARD_HANDLER = "de.davis.passwordmanager.extra.DASHBOARD_HANDLER"
    }
}